src/io_callbacks.zig

```diff
@@ -0,0 +1,463 @@
+                error.SignalInterrupt => continue,
+                // Wait for some completions and then try again:
+                // See https://github.com/axboe/liburing/issues/281 re: error.SystemResources.
+                // Be careful also that copy_cqes() will flush before entering to wait (it does):
+                // https://github.com/axboe/liburing/commit/35c199c48dfd54ad46b96e386882e7ac341314c5
+                error.CompletionQueueOvercommitted, error.SystemResources => {
+                    try self.flush_completions(true);
+                    continue;
+                },
+                else => return err,
+            };
+            self.submitted += self.queued;
+            self.queued = 0;
+            break;
+        }
+    }
+
+    fn get_sqe(self: *IO, completion: *Completion) void {
+        const sqe = self.ring.get_sqe() catch |err| switch (err) {
+            error.SubmissionQueueFull => {
+                completion.next = null;
+                self.unqueued.push(completion);
+                return;
+            },
+        };
+        self.queued += 1;
+        completion.complete_prep(sqe);
+    }
+
+    /// This struct holds the data needed for a single io_uring operation
+    pub const Completion = struct {
+        io: *IO,
+        result: i32 = undefined,
+        next: ?*Completion = undefined,
+        operation: Operation,
+        // This is one of the usecases for c_void outside of C code and as such c_void will
+        // be replaced with anyopaque eventually: https://github.com/ziglang/zig/issues/323
+        context: ?*c_void,
+        callback: fn (context: ?*c_void, completion: *Completion, result: *const c_void) callconv(.C) void,
+
+        fn prep(completion: *Completion) void {
+            completion.io.get_sqe(completion);
+        }
+
+        fn complete_prep(completion: *Completion, sqe: *io_uring_sqe) void {
+            switch (completion.operation) {
+                .accept => |*op| {
+                    linux.io_uring_prep_accept(
+                        sqe,
+                        op.socket,
+                        &op.address,
+                        &op.address_size,
+                        op.flags,
+                    );
+                },
+                .close => |*op| {
+                    linux.io_uring_prep_close(sqe, op.fd);
+                },
+                .openat => |*op| {
+                    linux.io_uring_prep_openat(
+                        sqe,
+                        op.fd,
+                        op.path,
+                        op.flags,
+                        op.mode,
+                    );
+                },
+            }
+            sqe.user_data = @ptrToInt(completion);
+        }
+
+        fn run(completion: *Completion) void {
+            switch (completion.operation) {
+                .accept => |op| {
+                    const result = result: {
```

Comment:
The labelled break expression here is fine, but is it possible to reduce this slightly by using an if expression directly instead?

Reply:
Already changed that to a less verbose version in the lasted commit. I personally am not a huge fan of 
```zig
const x = if (foo) blk: {
    // ...
} else blk: {
    // ...
};
```
as I think having a single block is more readable so I tend to opt for the single block version when the if isn't a single statement.

Reply:
> I personally am not a huge fan

Yeah, actually me too. I also prefer a single block. Happy to do what you think here.